modules/user_accounts/php/NDB_Form_user_accounts.class.inc

```diff
@@ -413,59 +418,87 @@ class NDB_Form_user_accounts extends NDB_Form
-        if ((!empty($values['NA_UserID']) || !empty($values['UserID'])) && empty($this->identifier)) {
-            // check username's uniqueness
-            $result = $DB->pselectOne("SELECT COUNT(*) FROM users WHERE UserID = :UID",
-                    array('UID' => $values['UserID'] )
-                    );
-
-            if (Utility::isErrorX($result)) {
-                return PEAR::raiseError("DB Error: ".$result->getMessage());
+        //============================================
+        //         Validate UserID and NA_UserID
+        //============================================
+        
+        if(empty($this->identifier)) {
+            // Clicked on "UID == email" and specified a UID
+            if(!empty($values['UserID']) && $values['NA_UserID'] == 'on') {
+                $errors['UserID_Group'] = 'You cannot enter a user name if you want it to match the email address';
             }
-
-            if ($result > 0) {
-                $errors['UserID_Group'] = 'The user name already exists';
+            // Not clicked on "UID == email" and not specified a UID
+            elseif(empty($values['UserID']) && $values['NA_UserID'] != 'on') {
+                $errors['UserID_Group'] = 'You must enter a user name or choose to make it match the email address';
             }
+            // Either specified a UID or clicked on "UID = email" with a non-empty email
+            elseif(!empty($values['UserID']) || ($values['NA_UserID'] == 'on' && $values['Email'])) { 
+        	       $effectiveUID = empty($values['UserID']) ? $values['Email'] : $values['UserID'];
+        	       
+                // check username's uniqueness
+                $result = $DB->pselectOne("SELECT COUNT(*) FROM users WHERE UserID = :UID",
+                    array('UID' => $effectiveUID)
+                );
+
+                if (Utility::isErrorX($result)) {
+                    return PEAR::raiseError("DB Error: ".$result->getMessage());
+                }
 
-            if (strlen($values['UserID']) > 255) {
-                $errors['UserID_Group'] = 'The user name must be less than 255 characters long';
-            }
+                if ($result > 0) {
+                    $errors['UserID_Group'] = 'The user name already exists';
+                }
 
+                if (strlen($effectiveUID) > 255) {
+                    $errors['UserID_Group'] = 'The user name must not exsceed 255 characters';
+                }
+            }
         }
-        if (!empty($this->identifier)) {
-            $pass = $DB->pselectOne("SELECT Password_md5 FROM users WHERE UserID = :UID",
-                    array('UID' => $this->identifier) );
+        
+        //==================================
+        //        Password validation
+        //==================================
+        if(!empty($this->identifier)) {
+            $pass = $DB->pselectOne("SELECT IFNULL(Password_md5, Password_hash) as Current_password FROM users WHERE UserID = :UID",
```

Comment:
This should be COALESCE, which is a more standard function thatn IFNULL, and Password_hash should be the priority (first argument), not Password_md5